driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java

```diff
@@ -130,41 +120,56 @@ public boolean isClosed() {
         return closed;
     }
 
-    ServerCursor getServerCursor() {
-        return cursor;
+    private boolean getAndSetClosed() {
+        boolean oldClosed;
+        synchronized (lock) {
+            oldClosed = closed;
+            closed = true;
+        }
+        return oldClosed;
+    }
+
+    private ServerCursor getAndSetCursor(final ServerCursor cursor) {
+        ServerCursor oldCursor;
+        synchronized (lock) {
+            oldCursor = this.cursor;
+            this.cursor = cursor;
+        }
+        return oldCursor;
     }
 
     private boolean limitReached() {
         return Math.abs(limit) != 0 && count >= Math.abs(limit);
     }
 
-    private void getMore(final SingleResultCallback<List<T>> callback) {
+    private void getMore(final ServerCursor cursor, final SingleResultCallback<List<T>> callback) {
         connectionSource.getConnection(new SingleResultCallback<AsyncConnection>() {
```

Comment:
There's still a race between next and killCursorOnClose.  Though unlikely, it could end up in a situation where there is a non-null cursor but the connectionSource's reference count is 0, and getConnection throws an exception as a result.

One way to fix this is for next() to atomically get the cursor and also retain a reference to the connectionSource, and for killCursorOnClose() to atomically get the cursor using the same lock.  The latter doesn't need to retain the connectionSource though, because it's the only one that would release it back to 0.